red4ext-sys/src/interop.rs

```diff
@@ -46,6 +46,91 @@ unsafe impl ExternType for CName {
     type Kind = cxx::kind::Trivial;
 }
 
+#[derive(Debug, Clone, Default, PartialEq, Eq)]
+#[repr(C)]
+pub struct ResRef(RaRef);
+
+impl ResRef {
+    pub fn new(path: &str) -> Self {
+        Self(RaRef::new(path))
+    }
+}
+
+unsafe impl ExternType for ResRef {
+    type Id = type_id!("RED4ext::ResRef");
+    type Kind = cxx::kind::Trivial;
+}
+
+#[derive(Debug, Clone, Default, PartialEq, Eq)]
+#[repr(C)]
+struct RaRef(ResourcePath);
+
+impl RaRef {
+    fn new(path: &str) -> Self {
+        Self(ResourcePath::new(path))
+    }
+}
+
+#[derive(Debug, Clone, Default, PartialEq, Eq)]
+#[repr(transparent)]
+pub struct ResourcePath {
+    hash: u64,
+}
+
+impl ResourcePath {
+    const MAX_LENGTH: usize = 216;
+
+    /// accepts non-sanitized path of any length,
+    /// but final sanitized path length must be equals or inferior to 216 bytes
+    /// otherwise it returns default empty `ResourcePath`
+    fn new(path: &str) -> Self {
+        if path.is_empty() {
+            return Self::default();
+        }
+        let sanitized = path
+            .trim_start_matches(|c| c == '\'' || c == '\"')
+            .trim_end_matches(|c| c == '\'' || c == '\"')
+            .trim_start_matches(|c| c == '/' || c == '\\')
+            .trim_end_matches(|c| c == '/' || c == '\\')
+            .to_ascii_lowercase()
+            .replace_consecutive(&['/', '\\'], b"\\");
+        if sanitized.as_bytes().len() > Self::MAX_LENGTH {
+            return Self::default();
+        }
+        Self {
+            hash: fnv1a64(&sanitized),
+        }
+    }
```

Comment:
what is this sanitization based on, e.g. why " is being trimmed?
one more question I'd have is: are absolute resource paths allowed?
if all paths are relative it'd probably make more sense to provide a simple builder that guarantees to construct a safe path from individual components instead:
```rs
struct ResourcePathBuilder<'a> {
  components: Vec<Cow<'a, str>>
}

impl <'a> ResourcePathBuilder<'a> {
  fn join(self, component: impl Into<Cow<'a, str>>) -> Self {
    self.components.push(component.into());
    self
  }

  fn build(self) -> ResourcePath {
    let rendered = self.components.join("/");
    ResourcePath { hash: fnv1a64(&rendered) }
  }
}

```

Reply:
could even go as far as providing a convenience macro
```rs
#[macro_export]
macro_rules! res_path {
    ($base:expr, /$lit:literal $($tt:tt)*) => {
        crate::res_path!($base.join($lit), $($tt)*)
    };
    ($base:expr, ) => {
        $base
    };
    ($lit:literal $($tt:tt)*) => {
        crate::res_path!(ResourcePathBuilder::default(), $($tt)*).build()
    };
}
```

Reply:
I like the builder pattern, I'll definitely check to add it. But I guess since `ResRef` / `RaRef` can be sent across the bridge, I should probably keep my constructor around anyway ? (or is the constructor never called in FFI context, the instance being built with another means ?).

so actually here I tried to duplicate RED4ext.SDK C++ [logic for ResourcePath](https://github.com/WopsS/RED4ext.SDK/blob/master/include/RED4ext/ResourcePath.hpp#L76-L128) to the best of my understanding.

wrapping `"` and `'` are stripped there, which I guess is to make no distinction between e.g. `"C:\somewhere"`, `'C:/somewhere'` and `C:\somewhere`.

another observation is cyberdoc depicts an [IsValid instance method](https://jac3km4.github.io/cyberdoc/#25765), which I thought at first allowed to construct a `ResRef` from an invalid path, but now I tend to think it actually does not but instead highlights the fact that this could be a valid path to a missing file.

Reply:
we don't need a constructor to pass it to/from the game - it works because the memory representation is the same so it can be simply passed through directly

I also think we don't need to follow the implementation directly, for example I think if there's a `new` method it should return either `Option` or `Result` to signal failure instead of returning a `default`

Reply:
ok that's what I wanted to know, then I'm gonna try to come up with a more idiomatic implementation in the coming days, thanks @jac3km4

Reply:
in my latest commit I did constructors-related changes but slightly differently, relying on `std::path::PathBuf`.
if you prefer, I can stick faithfully to your suggestion too @jac3km4.

Reply:
few suggestions:
- I think all resource paths are absolute starting from the root, without an explicit `/`, the first component seems to always be either `base` or `multi`, so I'd avoid using default Path types because they can be relative; if we use them we should validate that they're not relative, if you don't want to bother with that you can remove the builder
- I think the `replace_consecutive` logic can be simplified with a simple `split`:
```
   fn new(path: &str) -> Result<Self, ResourcePathError> {
        if path.is_empty() {
            return Err(ResourcePathError::Empty);
        }
        let sanitized = path
          .trim_start_matches(|c| c == '\'' || c == '\"')
          .trim_end_matches(|c| c == '\'' || c == '\"')
          .trim_start_matches(|c| c == '/' || c == '\\')
          .trim_end_matches(|c| c == '/' || c == '\\')
          .split(|c| c == '/' || c == '\\')
          .filter(|comp| !comp.is_empty())
          .map(|c| c.to_ascii_lowercase())
          .collect::<String>();
        if sanitized.as_bytes().len() > Self::MAX_LENGTH {
            return Err(ResourcePathError::TooLong {
                max: ResourcePath::MAX_LENGTH,
            });
        }
        Ok(Self {
            hash: fnv1a64(&sanitized),
        })
    }
```

Reply:
Ok I think I addressed all your concerns in 5913f49..833059c, let me know if I overlooked anything ^^